### DISCUSSION TITLE
Rollup issues with nut-js for sending key presses

### DISCUSSION THREAD

**Discussion by mrvonwyl** [elgatosf/elgato#102] [Help]
I am trying out the Stream Deck SDK with JavaScript.
For my project, I want to be able to send key presses. Basically like a macro.
The only maintained solution that I found is https://github.com/nut-tree/nut.js.
However, when I am adding this dependency Rollup is acting up and I am struggling to resolve it.

The issue seems to be jimp, which is a dependency of nut-js:
```
(!) Unresolved dependencies
https://rollupjs.org/troubleshooting/#warning-treating-module-as-external-dependency
jimp (imported by "jimp?commonjs-external")
```

When I follow the link, it says to use `nodeResolve()` which is already configured in the rollup.config.mjs.
Unfortunately, my Rollup knowledge is very limited.

Other ways to achieve key presses and ideally mouse clicks within a JavaScript plugin are appreciated as well.

**Comment by ChekTek** [comment]
I believe this is an issue with the module version of Jimp ([potentially related issue](https://github.com/jimp-dev/jimp/issues/1206)). 

I was able to get Jimp working by adding this to your `output` property in `rollup.config.mjs`:
```
inlineDynamicImports: true,
format: 'cjs',
```
and then changing the source to `commonjs` in the emitted file:
```
{
	name: 'emit-module-package-file',
	generateBundle() {
		this.emitFile({
			fileName: 'package.json',
			source: `{ "type": "commonjs" }`,
			type: 'asset',
		});
	},
},
```
and then importing Jimp like so
```
import Jimp from 'jimp';
```

Unfortunately, since the library you are using is depending on `Jimp`, rather than your project specifically, it may be very difficult to change the way it is imported, without the files getting overwritten during the build step.

**Comment by mrvonwyl** [comment]
Hi @ChekTek 

Thank you so much for the detailed solution. Unfortunately, I was not able to make it work.
But I am not sure If I understand your last paragraph correctly.
What I think you are saying is, you are able to use Jimp, but only standalone and not bundled inside @nut-tree/nut-js.

In any case, this is my full rollup.config.mjs. Maybe I made a mistake!?

```
import commonjs from "@rollup/plugin-commonjs";
import nodeResolve from "@rollup/plugin-node-resolve";
import terser from "@rollup/plugin-terser";
import typescript from "@rollup/plugin-typescript";
import path from "node:path";
import url from "node:url";
import Jimp from "jimp";

const isWatching = !!process.env.ROLLUP_WATCH;
const sdPlugin = "dev.elysion.scriptify.sdPlugin";

/**
 * @type {import('rollup').RollupOptions}
 */
const config = {
  input: "src/plugin.ts",
  output: {
    inlineDynamicImports: true,
    format: "cjs",
    file: `${sdPlugin}/bin/plugin.js`,
    sourcemap: isWatching,
    sourcemapPathTransform: (relativeSourcePath, sourcemapPath) => {
      return url.pathToFileURL(
        path.resolve(path.dirname(sourcemapPath), relativeSourcePath)
      ).href;
    },
  },
  plugins: [
    {
      name: "watch-externals",
      buildStart: function () {
        this.addWatchFile(`${sdPlugin}/manifest.json`);
      },
    },
    typescript({
      mapRoot: isWatching ? "./" : undefined,
    }),
    nodeResolve({
      browser: false,
      exportConditions: ["node"],
      preferBuiltins: true,
    }),
    commonjs(),
    !isWatching && terser(),
    {
      name: "emit-module-package-file",
      generateBundle() {
        this.emitFile({
          fileName: "package.json",
          source: `{ "type": "commonjs" }`,
          type: "asset",
        });
      },
    },
  ],
};

export default config;
```

**Reply by ChekTek** [reply]
Correct. I was able to use Jimp as a standalone library because I could decide how to import and use it, but because Jimp is used in the library, it may be difficult to get a configuration that work for both Jimp and nut-js.
